File: frontend/src/types/development.ts
// ============================================================
// TIPOS - DEVELOPMENT
// ============================================================
//
// Responsabilidade:
//     Centraliza os tipos TypeScript utilizados pela área de
//     Desenvolvimento do DUET CORE.
//
// Este arquivo descreve estruturas utilizadas em:
//     - projetos de Desenvolvimento;
//     - menus de projeto;
//     - origem de novos projetos;
//     - Agents para execução;
//     - Workflow / Kanban;
//     - Checkout;
//     - planejamento dos cards;
//     - comentários;
//     - Release.
//
// Este arquivo contém SOMENTE definições de tipos.
//
// Não deve:
//     - possuir estado React;
//     - realizar chamadas HTTP;
//     - conter regras de negócio;
//     - renderizar componentes.
//
// Objetivo arquitetural:
//     Evitar que Development.tsx concentre modelos de dados que
//     podem ser reutilizados por componentes, hooks e services.
// ============================================================


// ============================================================
// MENU DO PROJETO
// ============================================================

export interface ProjectMenuState {

    // Projeto ao qual o menu aberto pertence.
    projectId: number;

    // Coordenadas da tela utilizadas pelo menu renderizado
    // através de Portal.
    top: number;
    left: number;
}


// ============================================================
// PROJETO DE DESENVOLVIMENTO
// ============================================================
//
// Representa um AutomationProject persistido no Control Room.
//
// O mesmo projeto pode aparecer:
//     - na lista tradicional;
//     - no Workflow / Kanban;
//     - no DUET Studio.
// ============================================================

export interface DevelopmentProject {

    // ID real gerado pelo PostgreSQL.
    id: number;

    // Dados principais do projeto.
    name: string;
    description: string | null;

    // Pasta da área de Desenvolvimento.
    folder_id: number | null;

    // Estado técnico do projeto.
    status:
        | "draft"
        | "modified"
        | "published";

    // Estágio atual do Workflow.
    current_stage_id: number | null;

    // Origem em Produção, quando o projeto nasceu
    // de um Robot já publicado.
    base_robot_id: number | null;
    // Nome amigável do Robot publicado que originou o projeto.
    //
    // null:
    //     projeto criado do zero ou Robot de origem indisponível.
    base_robot_name: string | null;
    base_version: number | null;


    // ========================================================
    // AUDITORIA
    // ========================================================

    created_by: number;
    created_at: string | null;
    updated_at: string | null;


    // ========================================================
    // CHECKOUT ATUAL
    // ========================================================
    //
    // null:
    //     ninguém está editando o projeto.
    //
    // objeto:
    //     identifica o usuário que possui o Checkout exclusivo.
    // ========================================================

    checkout?: {
        id: number;
        project_id: number;
        user_id: number;
        user_name: string | null;
        checked_out_at: string | null;
    } | null;


    // ========================================================
    // PLANEJAMENTO DO CARD
    // ========================================================
    //
    // Estes dados pertencem individualmente ao projeto.
    //
    // Responsável técnico NÃO representa o usuário que está
    // atualmente em Checkout.
    // ========================================================

    functional_responsible_id: number | null;
    functional_responsible_name?: string | null;

    technical_responsible_id: number | null;
    technical_responsible_name?: string | null;

    start_date: string | null;
    due_date: string | null;

    effort_hours: number | null;

    // O board devolve somente a quantidade.
    // Os comentários completos são carregados quando o painel
    // de detalhes do card é aberto.
    comments_count?: number;


    // ========================================================
    // LIXEIRA
    // ========================================================

    deleted_at: string | null;
    deleted_by: number | null;
    deleted_by_name?: string | null;


    // Soft delete.
    is_active: boolean;
}


// ============================================================
// ORIGEM DO NOVO PROJETO
// ============================================================
//
// "new":
//     automação criada do zero.
//
// "existing":
//     projeto criado a partir de um Robot existente.
// ============================================================

export type ProjectOriginMode =
    | "new"
    | "existing";


export interface OriginRobotFolder {

    // ID real da pasta na estrutura de Robôs.
    id: number;

    // Nome apresentado ao usuário.
    name: string;

    // null representa uma pasta diretamente na raiz.
    parent_id: number | null;
}


export interface OriginRobot {

    // Identidade permanente do Robot.
    id: number;

    // Nome atual em Produção.
    name: string;

    // Versão atualmente publicada.
    version: number;

    // Pasta real do Robot.
    folder_id: number | null;

    // Caminho amigável utilizado pela interface.
    folder_label: string;
}


// ============================================================
// AGENTS PARA EXECUÇÃO
// ============================================================
//
// Estrutura retornada pelo endpoint:
//     /agents/execution/available-agents
//
// Contém somente os dados necessários para o usuário escolher
// onde executar o projeto de Desenvolvimento.
// ============================================================

export interface ExecutionAgent {

    agent_id: string;

    name: string;

    host: string;
    port: number;

    status: string;

    session_status: string | null;

    username: string | null;
}


// ============================================================
// RELEASE
// ============================================================
//
// Representa a prévia calculada pelo backend antes de publicar
// um projeto de Desenvolvimento em Produção.
// ============================================================

export interface ReleasePreview {

    // Token que garante que a publicação utilize exatamente
    // a prévia que foi calculada.
    preview_token: string;


    // Robot existente quando o projeto representa uma nova
    // versão de uma automação já publicada.
    robot: {
        id: number;
        name: string;
        version: number;
        folder_id: number | null;
    } | null;


    // Próxima versão do Robot.
    next_version: number;


    // Pacotes encontrados no projeto que podem ser convertidos
    // em Libraries durante o Release.
    library_candidates: string[];


    // Estrutura de pastas de Produção disponível para destino.
    folders: {
        id: number;
        name: string;
        parent_id: number | null;
    }[];


    // Libraries já vinculadas ao projeto.
    dependencies: {
        library_id: number;
        library_name: string;
        version: string;
        modified: boolean;
        suggested_version: string;
    }[];
}


// ============================================================
// WORKFLOW / KANBAN
// ============================================================

export type DevelopmentViewMode =
    | "projects"
    | "kanban";


export interface DevelopmentStage {

    // Identificação do estágio persistido.
    id: number;

    code: string;
    name: string;

    position: number;

    is_active: boolean;


    // Quantidade total de projetos da coluna.
    total_projects: number;


    // Projetos pertencentes ao estágio.
    projects: DevelopmentProject[];
}


// ============================================================
// DETALHES DO CARD
// ============================================================

export interface CardUser {

    // Usuário ativo disponível para responsabilidade
    // funcional ou técnica.
    id: number;

    name: string;
}


export interface CardComment {

    // Identidade do comentário.
    id: number;

    // Projeto ao qual pertence.
    project_id: number;


    // Autor obtido pela sessão autenticada.
    user_id: number;

    user_name: string | null;


    // Conteúdo do comentário.
    content: string;


    // Auditoria.
    created_at: string | null;

    updated_at: string | null;
}

// ============================================================
// PAYLOAD - CRIAÇÃO DE PROJETO
// ============================================================
//
// Estrutura enviada ao backend ao criar um novo
// AutomationProject na área de Desenvolvimento.
// ============================================================

export interface CreateDevelopmentProjectPayload {

    // Título da demanda.
    name: string;

    // Descrição opcional.
    description: string | null;

    // Pasta de Desenvolvimento.
    folder_id: number | null;

    // Robot de Produção utilizado como origem.
    //
    // null:
    //     projeto criado do zero.
    base_robot_id: number | null;
}


// ============================================================
// PAYLOAD - PLANEJAMENTO DO CARD
// ============================================================
//
// Estrutura enviada ao endpoint de atualização dos metadados
// de planejamento de um card do Kanban.
// ============================================================

export interface CardDetailsUpdatePayload {

    functional_responsible_id: number | null;

    technical_responsible_id: number | null;

    start_date: string | null;

    due_date: string | null;

    effort_hours: number | null;
}


// ============================================================
// PAYLOAD - NOVA LIBRARY DURANTE RELEASE
// ============================================================
//
// Representa um pacote do projeto que será publicado como uma
// nova Library durante o processo de Release.
// ============================================================

export interface NewReleaseLibraryPayload {

    import_name: string;

    name: string;

    version: string;
}


// ============================================================
// PAYLOAD - PUBLICAÇÃO
// ============================================================
//
// Estrutura enviada ao backend quando o usuário confirma o
// Release de um AutomationProject.
//
// Ela mantém:
//     - confirmação da demanda;
//     - token da prévia;
//     - destino do Robot;
//     - nome definitivo de Produção;
//     - versões das Libraries;
//     - novas Libraries criadas durante o Release.
// ============================================================

export interface PublishDevelopmentProjectPayload {

    confirmation_name: string;

    preview_token: string;

    folder_id: number | null;

    new_folder_path: string;

    robot_name: string;

    library_versions: Record<number, string>;

    new_libraries: NewReleaseLibraryPayload[];
}

File: schemas/development.py
# ============================================================
# SCHEMAS - DEVELOPMENT
# ============================================================
#
# Contratos de entrada utilizados pela API de Desenvolvimento
# do DUET CORE.
#
# Este módulo contém somente estruturas Pydantic.
#
# Não possui:
#
# - acesso ao banco;
# - regras de negócio;
# - acesso ao filesystem;
# - autenticação;
# - RBAC;
# - endpoints FastAPI.
#
# Dessa forma, os mesmos contratos podem ser utilizados pelo
# router e pelos services sem criar dependência da camada HTTP.
# ============================================================


from datetime import date

from pydantic import BaseModel, Field

# O contrato de publicação já pertence ao domínio de Release.
#
# Development apenas reutiliza esse contrato para manter
# compatibilidade com a API atual.
from releases.service import ReleaseRequest


# ============================================================
# PASTAS
# ============================================================

class DevelopmentFolderCreate(BaseModel):
    """
    Dados necessários para criar uma pasta na área
    de Desenvolvimento.

    parent_id:
        None:
            cria uma pasta na raiz.

        ID preenchido:
            cria uma subpasta dentro da pasta informada.
    """

    name: str = Field(
        ...,
        min_length=1,
        max_length=255,
    )

    parent_id: int | None = None


# ============================================================
# PROJETOS
# ============================================================

class AutomationProjectCreate(BaseModel):
    """
    Dados necessários para criar um projeto de automação.

    folder_id:
        Pasta da área de Desenvolvimento onde o projeto
        será criado.

        Pode ser None para projetos na raiz.

    base_robot_id:
        Quando preenchido, indica que o projeto nasceu
        a partir de um Robot já publicado.

        A versão atual do Robot é capturada pelo backend
        e armazenada como base_version.

        Para automações novas deve permanecer None.
    """

    name: str = Field(
        ...,
        min_length=1,
        max_length=255,
    )

    description: str | None = Field(
        default=None,
        max_length=5000,
    )

    folder_id: int | None = None

    base_robot_id: int | None = None


# ============================================================
# DETALHES DO CARD / KANBAN
# ============================================================

class AutomationProjectCardUpdate(BaseModel):
    """
    Atualiza os dados de planejamento pertencentes
    individualmente a um card/projeto.

    Todos os campos são opcionais porque:

    - projetos antigos podem não possuir planejamento;
    - um campo pode ser definido posteriormente;
    - enviar explicitamente None permite limpar o valor.
    """

    # Usuário do DUET responsável pela parte funcional.
    functional_responsible_id: int | None = None

    # Usuário do DUET responsável pela parte técnica.
    #
    # Este campo NÃO representa quem possui Checkout.
    technical_responsible_id: int | None = None

    # Data de início do trabalho.
    start_date: date | None = None

    # Previsão de conclusão.
    due_date: date | None = None

    # Horas de esforço deste card.
    #
    # Não existe cronômetro.
    # É apenas um valor informado manualmente.
    effort_hours: float | None = Field(
        default=None,
        ge=0,
        le=999999.99,
    )


class ProjectCommentCreate(BaseModel):
    """
    Dados necessários para adicionar um comentário
    livre ao card do Kanban.
    """

    content: str = Field(
        ...,
        min_length=1,
        max_length=5000,
    )


# ============================================================
# WORKFLOW / KANBAN
# ============================================================

class DevelopmentStageMove(BaseModel):
    """
    Dados necessários para movimentar um projeto entre
    estágios normais do Workflow.

    target_stage_id:
        ID do estágio de destino obtido através da API:

            /development/workflow/stages

    IMPORTANTE:
        O estágio PUBLISHED não pode ser alcançado através
        da movimentação normal.

        Publicação possui fluxo protegido próprio.
    """

    target_stage_id: int


# ============================================================
# RELEASE
# ============================================================

# Mantemos exatamente o nome utilizado atualmente pela API
# de Development.
#
# O contrato real continua pertencendo ao release_service,
# evitando duplicação do schema de publicação.
DevelopmentPublishRequest = ReleaseRequest


# ============================================================
# WORKSPACE - SALVAR ARQUIVO
# ============================================================

class WorkspaceFileSave(BaseModel):
    """
    Dados utilizados para salvar um arquivo existente
    dentro do workspace de um AutomationProject.

    path:
        Caminho relativo dentro do workspace.

        Exemplos:

            main.py
            src/processamento.py

    content:
        Conteúdo textual completo que será salvo.
    """

    path: str = Field(
        ...,
        min_length=1,
        max_length=1000,
    )

    # O Studio é um editor de código/configuração textual.
    #
    # O limite evita que um único request faça o backend
    # materializar conteúdo textual arbitrariamente grande.
    # 5 MiB de caracteres é muito acima do necessário para
    # código-fonte normal e ainda preserva arquivos maiores.
    content: str = Field(
        ...,
        max_length=5 * 1024 * 1024,
    )


# ============================================================
# WORKSPACE - CRIAR ITEM
# ============================================================

class WorkspaceItemCreate(BaseModel):
    """
    Utilizado para criar arquivos ou pastas dentro
    de um workspace.

    O path é sempre relativo à raiz do projeto.

    Exemplos:

        utils.py
        src
        src/services
        src/processamento.py
    """

    path: str = Field(
        ...,
        min_length=1,
        max_length=1000,
    )


# ============================================================
# WORKSPACE - RENOMEAR ITEM
# ============================================================

class WorkspaceItemRename(BaseModel):
    """
    Dados necessários para renomear um arquivo ou pasta
    dentro do workspace.

    path:
        Caminho atual relativo ao workspace.

        Exemplos:

            teste.py
            src/utils.py
            elements/botoes

    new_name:
        Novo nome do arquivo ou pasta.

    IMPORTANTE:
        O rename não move o item para outra pasta.

        Ele altera somente o nome dentro do mesmo diretório.
    """

    path: str = Field(
        ...,
        min_length=1,
        max_length=1000,
    )

    new_name: str = Field(
        ...,
        min_length=1,
        max_length=255,
    )


# ============================================================
# BIBLIOTECA NOVA DO PROJETO
# ============================================================

class DevelopmentLibraryCreate(BaseModel):
    """
    Cria uma biblioteca reutilizável dentro do projeto atual.

    A Library nasce como identidade reservada, porém ainda
    NÃO aparece no catálogo de Produção.

    Sua primeira versão será criada somente quando o projeto
    for publicado.
    """

    name: str = Field(
        ...,
        min_length=1,
        max_length=255,
    )

    import_name: str = Field(
        ...,
        min_length=1,
        max_length=255,
    )

    description: str | None = Field(
        default=None,
        max_length=5000,
    )